services/multilang-support/src/utils.py

```diff
@@ -0,0 +1,221 @@
+            if device_id >= num_gpus:
+                LOGGER.warning("No GPU available, using CPU.")
+                set_device(device_id=-1)
+                break
+    LOGGER.debug(f"Model is running on {DEVICE_NAME}.")
+
+
+@log
+def tokenize_query_to_tensor(query: str):
+    """
+    STEP 1: Tokenize the query using the tokenizer.
+    @param query: The query (type<str>) to be tokenized.
+    @return: The tokenized query as a torch.Tensor.
+    """
+    # return TOKENIZER.encode(query).to(DEVICE)
+    return TOKENIZER(query, return_tensors="pt").to(DEVICE).input_ids
+
+
+@log
+def generate_output_tensor(input_ids: torch.Tensor) -> torch.Tensor:
+    """
+    STEP 2: Translate the input_ids tensor to an output query.
+    @param input_ids: The decoded tensor (type<torch.Tensor>) to be translated.
+    @return: Newly generated tensor using the model (type<torch.Tensor>).
+    TODO: Add parameters to control/optimize the translation.
+    """
+    return MODEL.generate(input_ids, max_time=MAX_TIME)
+
+
+@log
+def decode_generated_tensor(translated_tensor: torch.Tensor) -> str:
+    """
+    STEP 3: Decode the translated tensor to a string.
+    @param translated_tensor: <class 'torch.Tensor'>
+    @return: <class 'str'>
+    """
+    translated_result = TOKENIZER.decode(
+        translated_tensor[0],
+        skip_special_tokens=True,
+        clean_up_tokenization_spaces=True,
+    )
+    return translated_result
+
+
+@log
+def translate_helsinki(segment: list) -> list:
+    """
+    Translate the segment - (a list of strings) - using Helsinki-NLP models.
+    Steps:
+    @param segment: The segment (type<list>) to be translated.
+    1. Tokenize the segment to a tensor.
+    2. Pass tokenized tensor through the model.
+    3. Decode the translated tensor to a string.
+    4. Append the translated string to a list.
+    """
+    # output list
+    result = []
+    for input_query in segment:
+        # 1. Input query -> tensor
+        LOGGER.debug(f'(1) Tokenizing input segment "{input_query}"')
+        input_tensor, _time_inTensor = tokenize_query_to_tensor(input_query)
+
+        # 2. Input tensor -> output tensor
+        LOGGER.debug("(2) Generating new tensor.")
+        output_tensor, _time_outTensor = generate_output_tensor(input_tensor)
+
+        # 3. Output tensor -> query
+        LOGGER.debug("(3) Decoding translated tensor.")
+        output_query, _time_outQuery = decode_generated_tensor(output_tensor)
+
+        # 4. Translated query -> result
+        LOGGER.debug(f'(4) Appending "{output_query}" to result.')
+        LOGGER.info(f'Translated: "{output_query}"')
+        result.append(output_query)
+
+    return result
+
+
+if "utils" in __name__ or __name__ == "__main__":
+    instantiate()
+    LOGGER.debug(f"Using {DEVICE} on {DEVICE_NAME}")
+    LOGGER.info('Service is instantiated and ready.')
+
+
+class Translator:
+    """
+    (Not yet implemented fully)
```

Comment:
What is remaining for this to be implemented?

Reply:
My idea is to generalise the service into `Translator` objects for each language pair. Each object will perform tasks like the service right now using their own `MODEL` and `TOKENIZER` instead of a global one. 

Basically, I (or future developers) will have to implement all the functions above but as a class function (with e.g., `self.MODEL`) and calling `Translator('en', 'es').translate(segments)` in translate.py

Reply:
I think this part can be ignored to finish the sprint/PR. I had a discussion with @rbazin  about ML and using the models, and we figured that there's another approach to implementing the service that is faster and utilize the resources better, since it uses the Marian framework that it was trained on rather than via an API like HuggingFace or Pytorch.